pkg/js/util/exceptions.ts

```diff
@@ -0,0 +1,117 @@
+    SelfError = "self-error",
+    InvalidName = "invalid-name",
+    MissingDefinition = "missing-definition",
+    InvalidRelationType = "invalid-relation-type",
+    InvalidType = "invalid-type",
+    RelationNoEntrypoint = "relation-no-entry-point",
+    TuplesetNotDirect = "tupleuset-not-direct",
+    DuplicatedError = "duplicated-error",
+    RequireSchema1_0 = "allowed-type-schema-10",
+    AssignableRelationsMustHaveType = "assignable-relation-must-have-type",
+    AllowedTypesNotValidOnSchema1_0 = "allowed-type-not-valid-on-schema-1_0",
+    InvalidSchema = "invalid-schema",
+    InvalidSyntax = "invalid-syntax",
+    TypeRestrictionCannotHaveWildcardAndRelation = "type-wildcard-relation",
+}
+
+/**
+ * Abstract base class for syntax and validation exceptions
+ */ 
+abstract class OpenFGABaseError extends Error {
+    constructor(
+        public line: number,
+        public column: number,
+        public msg: string,
+        public type: string,
+    ) {
+        super(`${type} error at line=${line}, column=${column}: ${msg}`);
+    }
+
+    toString() {
+        return this.message;
+    }
+}
+
+/** 
+ * Added to listener during syntax parsing, when syntax errors are encountered
+ */ 
+export class OpenFgaDslSyntaxError extends OpenFGABaseError {
+    constructor(
+        public line: number,
+        public column: number,
+        public msg: string,
+        public metadata?: {
+            symbol: string,
+            start: number,
+            stop: number,
+        },
+        e?: RecognitionException,
+    ) {
+        super(line, column, msg, "syntax");
+        if (e?.stack) {
+            this.stack = e.stack;
+        }
+        this.metadata = metadata;
+    }
+
+    toString() {
+        return this.message;
+    }
+}
+
+/**
+ * Thrown at the end of syntax parsing, collecting all Syntax errors encountered during parsing
+ */
+export class OpenFgaDslSyntaxMultipleError extends Error {
+    constructor(public errors: OpenFgaDslSyntaxError[]) {
+        super(`${errors.length} error${errors.length > 1 ? "s" : ""} occurred:\n\t* ${errors.join("\n\t* ")}\n\n`);
+        this.errors = errors;
+    }
+
+    toString() {
+        return this.message;
+    }
+}
+
+/**
+ * Added to reporter as the JSON transformation is being parsed and validated
+ */
+export class OpenFgaJsonValidationError extends OpenFGABaseError {
+    constructor(
+        public line: number,
+        public column: number,
+        public msg: string,
+        public metadata?: {
+            symbol: string,
+            errorType: ValidationError,
+        },
+    ) {
+        super(line, column, msg, metadata?.errorType.toString() || "validation");
+        this.metadata = metadata;
+    }
+
+    toString() {
+        return this.message;
+    }
+}
+
+/** 
+ * Thrown at end of checkDSL validation, collecting all encountered validation errors
+ */
+export class OpenFgaValidationMultipleError extends Error {
```

Comment:
Re error class naming, consider using:

- `DSLSyntaxError` for the DSL syntax errors
- `ModelValidationError` for the JSON validation errors,
- `DSLValidationError`, for the wrapped `ModelValidationError` that adds the DSL metadata and conforms to the same output as the `DSLSyntaxError`

Just to be consistent with errors elsewhere, call the top level interface that (instead of multiple).

The individual errors can be called:
`DSLSyntaxSingleError` etc..

Reply:
Make sure the classes for these errors are exported from `pkg/js/errors` as consumers will probably be using them